### ISSUE TITLE
Geckodriver log level does not work

### ISSUE THREAD

**Issue by atomar09** [mozilla/geckodriver#1317]
## System

* Version: 0.20.1
* Platform: Mac OSX 10.13.5
* Firefox: 59.0.2 64-bit
* Selenium: 3.12

## Testcase
No HTML code is required for this
Even if we use FATAL or ERROR for FirefoxDriverLogLevel, it still prints the INFO log messages.
FirefoxOptions options = new FirefoxOptions();
options.setLogLevel(FirefoxDriverLogLevel.FATAL);

Also, is there any way to turn off the logs?

## Stacktrace
This is what was printed with FATAL log level

```
1530216248944	geckodriver	INFO	geckodriver 0.20.1
1530216248952	geckodriver	INFO	Listening on 127.0.0.1:44099
2018-06-28 13:04:10.363 plugin-container[1688:96703] *** CFMessagePort: bootstrap_register(): failed 1100 (0x44c) 'Permission denied', port = 0x1503f, name = 'com.apple.tsm.portname'
See /usr/include/servers/bootstrap_defs.h for the error codes.
Jun 28, 2018 1:04:10 PM org.openqa.selenium.remote.ProtocolHandshake createSession
INFO: Detected dialect: W3C
2018-06-28 13:04:11.284 plugin-container[1689:96814] *** CFMessagePort: bootstrap_register(): failed 1100 (0x44c) 'Permission denied', port = 0xaa07, name = 'com.apple.tsm.portname'
See /usr/include/servers/bootstrap_defs.h for the error codes.
```

## Trace-level log

```
1530216290560	geckodriver	INFO	geckodriver 0.20.1
1530216290569	geckodriver	INFO	Listening on 127.0.0.1:31902
1530216290804	mozrunner::runner	INFO	Running command: "/Applications/Firefox.app/Contents/MacOS/firefox-bin" "-marionette" "-profile" "/var/folders/hr/lf3xj9js0_1gtld_ldw7jynh32drgp/T/rust_mozprofile.VwJROUBmO8T1"
1530216290806	geckodriver::marionette	DEBUG	Waiting 60s to connect to browser
1530216291128	Marionette	DEBUG	Received observer notification "profile-after-change"
1530216291230	Marionette	DEBUG	Received observer notification "command-line-startup"
1530216291230	Marionette	INFO	Enabled via --marionette
2018-06-28 13:04:51.968 plugin-container[1693:97876] *** CFMessagePort: bootstrap_register(): failed 1100 (0x44c) 'Permission denied', port = 0x14f3f, name = 'com.apple.tsm.portname'
See /usr/include/servers/bootstrap_defs.h for the error codes.
1530216292128	Marionette	DEBUG	Received observer notification "sessionstore-windows-restored"
1530216292284	Marionette	DEBUG	Setting recommended pref browser.newtabpage.enabled to false
1530216292284	Marionette	DEBUG	Setting recommended pref browser.newtabpage.introShown to true
1530216292284	Marionette	DEBUG	Setting recommended pref toolkit.cosmeticAnimations.enabled to false
1530216292284	Marionette	DEBUG	Setting recommended pref datareporting.policy.dataSubmissionPolicyAccepted to false
1530216292284	Marionette	DEBUG	New connections are accepted
1530216292286	Marionette	INFO	Listening on port 53012
1530216292354	geckodriver::marionette	DEBUG	Connected to Marionette on localhost:53012
1530216292354	Marionette	DEBUG	Accepted connection 0 from 127.0.0.1:53049
1530216292356	geckodriver::marionette	TRACE	<- {"applicationType":"gecko","marionetteProtocol":3}
1530216292356	geckodriver::marionette	TRACE	-> 163:[0,1,"newSession",{"acceptInsecureCerts":true,"browserName":"firefox","capabilities":{"desiredCapabilities":{"acceptInsecureCerts":true,"browserName":"firefox"}}}]
1530216292356	Marionette	TRACE	0 -> [0,1,"newSession",{"acceptInsecureCerts":true,"browserName":"firefox","capabilities":{"desiredCapabilities":{"acceptInsecureCerts":true,"browserName":"firefox"}}}]
1530216292356	Marionette	WARN	TLS certificate errors will be ignored for this session
1530216292414	Marionette	DEBUG	Register listener.js for window 2147483649
1530216292420	Marionette	TRACE	0 <- [1,1,null,{"sessionId":"315c8a19-0e32-fc44-b488-d34d5f8ce9ce","capabilities":{"browserName":"firefox","browserVersion":"59.0. ... gtld_ldw7jynh32drgp/T/rust_mozprofile.VwJROUBmO8T1","moz:useNonSpecCompliantPointerOrigin":false,"moz:webdriverClick":true}}]
1530216292421	geckodriver::marionette	TRACE	<- [1,1,null,{"sessionId":"315c8a19-0e32-fc44-b488-d34d5f8ce9ce","capabilities":{"browserName":"firefox","browserVersion":"59.0.2","platformName":"darwin","platformVersion":"17.6.0","pageLoadStrategy":"normal","acceptInsecureCerts":true,"timeouts":{"implicit":0,"pageLoad":300000,"script":30000},"rotatable":false,"moz:accessibilityChecks":false,"moz:headless":false,"moz:processID":1692,"moz:profile":"/var/folders/hr/lf3xj9js0_1gtld_ldw7jynh32drgp/T/rust_mozprofile.VwJROUBmO8T1","moz:useNonSpecCompliantPointerOrigin":false,"moz:webdriverClick":true}}]
1530216292421	webdriver::server	DEBUG	<- 200 OK {"value": {"sessionId":"315c8a19-0e32-fc44-b488-d34d5f8ce9ce","capabilities":{"acceptInsecureCerts":true,"browserName":"firefox","browserVersion":"59.0.2","moz:accessibilityChecks":false,"moz:headless":false,"moz:processID":1692,"moz:profile":"/var/folders/hr/lf3xj9js0_1gtld_ldw7jynh32drgp/T/rust_mozprofile.VwJROUBmO8T1","moz:useNonSpecCompliantPointerOrigin":false,"moz:webdriverClick":true,"pageLoadStrategy":"normal","platformName":"darwin","platformVersion":"17.6.0","rotatable":false,"timeouts":{"implicit":0,"pageLoad":300000,"script":30000}}}}
Jun 28, 2018 1:04:52 PM org.openqa.selenium.remote.ProtocolHandshake createSession
INFO: Detected dialect: W3C
1530216292460	webdriver::server	DEBUG	-> POST /session/315c8a19-0e32-fc44-b488-d34d5f8ce9ce/timeouts {"implicit":3000}
1530216292460	geckodriver::marionette	TRACE	-> 37:[0,2,"setTimeouts",{"implicit":3000}]
1530216292462	Marionette	TRACE	0 -> [0,2,"setTimeouts",{"implicit":3000}]
1530216292462	Marionette	TRACE	0 <- [1,2,null,{}]
1530216292464	geckodriver::marionette	TRACE	<- [1,2,null,{}]
1530216292464	webdriver::server	DEBUG	<- 200 OK {"value": {}}
1530216292466	webdriver::server	DEBUG	-> POST /session/315c8a19-0e32-fc44-b488-d34d5f8ce9ce/url {"url":"https://www.google.com"}
1530216292466	geckodriver::marionette	TRACE	-> 44:[0,3,"get",{"url":"https://www.google.com"}]
1530216292466	Marionette	TRACE	0 -> [0,3,"get",{"url":"https://www.google.com"}]
1530216292470	Marionette	DEBUG	Received DOM event beforeunload for about:blank
1530216292804	Marionette	DEBUG	Received DOM event pagehide for about:blank
1530216292892	Marionette	DEBUG	Received DOM event DOMContentLoaded for https://www.google.com/
2018-06-28 13:04:52.894 plugin-container[1694:97963] *** CFMessagePort: bootstrap_register(): failed 1100 (0x44c) 'Permission denied', port = 0xaa43, name = 'com.apple.tsm.portname'
See /usr/include/servers/bootstrap_defs.h for the error codes.
1530216293086	Marionette	DEBUG	Received DOM event pageshow for https://www.google.com/
1530216293088	Marionette	TRACE	0 <- [1,3,null,{}]
1530216293090	geckodriver::marionette	TRACE	<- [1,3,null,{}]
1530216293090	webdriver::server	DEBUG	<- 200 OK {"value": {}}
```

**Comment by whimboo** [comment]
So the log even shows trace level entries which should not appear by default. So I assume the code-snippet as shown above is not as what has been used to gather this log? Also which binding are you using? 

I tried myself via the Python bindings and by using "error". I indeed see INFO lines but no DEBUG and definitely no TRACE lines.

> from selenium.webdriver.firefox.options import Options
> opts = Options()
> opts.log.level = "trace"

**Comment by whimboo** [comment]
Also please note that there is issue #1046 for the Firefox logs.

**Comment by atomar09** [comment]
The stacktrace is captured with log level FATAL and it still showed INFO logs.

The bugs asks to provide TRACE logs, this is why I changed the log level to trace to provide TRACE logs.
JAVA binding is used.

As you also saw that INFO lines show up even if you use ERROR in Python binding also.
Why INFO is showing, it is less than ERROR and FATAL?

What is the way to turn off all the logs?

**Comment by andreastt** [comment]
When you use capabilities to configure the log level, these are parsed at session creation. The INFO lines quoted above are emitted when geckodriver starts, which is _before_ a new session is created. After the session has been started, you will not see any further INFO-level log messages.

To not emit the first INFO-level lines you need to pass `--log-level fatal` as a command-line flag to geckodriver when starting it, as the `moz:firefoxOptions` capability only controls the logging configuration of the lifetime of the session.

In conclusion, this is expected behaviour.

**Comment by andreastt** [comment]
What we could consider doing is to return the geckodriver version as part of the capabilities (`moz:geckodriverVersion` perhaps?) and drop the second `Listening on …` line altogether as I don’t think it adds much value.

```
1530216248944	geckodriver	INFO	geckodriver 0.20.1
1530216248952	geckodriver	INFO	Listening on 127.0.0.1:44099
```

I’ve filed https://bugzilla.mozilla.org/show_bug.cgi?id=1477474 to explore these options.

**Comment by atomar09** [comment]
@andreastt , I think that makes sense.
Could you please also let me know how to remove these lines?
 plugin-container[1688:96703] *** CFMessagePort: bootstrap_register():

**Comment by atomar09** [comment]
Could you please point me to Java and Python examples for this?

**Comment by andreastt** [comment]
I don’t, but it’s essentially about redirecting the stdout and stderr of the subprocess, which is exposed in both language’s standard libraries. I don’t know if Selenium exposes it directly, but if they don’t, filing an issue on having a choice on whether the driver’s stdout should be propagated seems like a good idea.

**Comment by atomar09** [comment]
@andreastt 
Do you know why these lines are showing up?
From where they are showing up?
2018-06-28 13:04:10.363 plugin-container[1688:96703] *** CFMessagePort: bootstrap_register(): failed 1100 (0x44c) 'Permission denied', port = 0x1503f, name = 'com.apple.tsm.portname'
See /usr/include/servers/bootstrap_defs.h for the error codes.
Jun 28, 2018 1:04:10 PM org.openqa.selenium.remote.ProtocolHandshake createSession
INFO: Detected dialect: W3C
2018-06-28 13:04:11.284 plugin-container[1689:96814] *** CFMessagePort: bootstrap_register(): failed 1100 (0x44c) 'Permission denied', port = 0xaa07, name = 'com.apple.tsm.portname'
See /usr/include/servers/bootstrap_defs.h for the error codes.

**Comment by andreastt** [comment]
@atomar09 Yes, these are mostly stdout lines from Firefox itself. The `INFO: Detected dialect: W3C` line is from the Selenium Java library.

**Comment by andreastt** [comment]
I just want to remind about https://bugzilla.mozilla.org/show_bug.cgi?id=1399441, which tracks hiding Firefox stdout output.

**Comment by atomar09** [comment]
@andreastt thanks for the information.
Is there a way to turn off stdout lines from Firefox?
